leetcode/279_perfect_squares.py

```diff
@@ -0,0 +1,27 @@
+class Solution:
+    def numSquares(self, n: int) -> int:
+        """
+        Finds the minimum number of perfect square numbers that sum up to n
+
+        Args:
+            n (int): The target number.
+
+        Returns:
+            int: The minimum number of perfect squares that sum up to n.
+        """
+        squares = []
+        i = 1
+        while i * i <= n:
+            squares.append(i * i)
+            i += 1
+
+        dp = [n] * (n + 1)
```

Comment:
This is very good. DP totally works here! Here your `dp` array acts like a visited yet. 

This problem really is closer to a BFS where you find the shortest path from 0 to n where there is an edge from `i -> j` if `i + x^2 = j` 

It's a lot of overlap between DP and BFS because the visited set is that part storing the intermediate results.

Reply:
No need for changes, just a comment here